lua/nvim-treesitter.lua

```diff
@@ -27,30 +27,58 @@ function M.define_modules(...)
   configs.define_modules(...)
 end
 
-function M.statusline(indicator_size)
+local get_line_for_node = function(node, type_patterns, transform_fn)
+  local node_type = node:type()
+  local is_valid = false
+  for _, rgx in ipairs(type_patterns) do
+    if node_type:find(rgx) then
+      is_valid = true
+      break
+    end
+  end
+  if not is_valid then return '' end
+  local line = transform_fn(vim.trim(ts_utils.get_node_text(node)[1] or ''))
+  -- Escape % to avoid statusline to evaluate content as expression
+  return line:gsub('%%', '%%%%')
+end
+
+-- Trim spaces and opening brackets from end
+local transform_line = function(line)
+  return line:gsub('%s*[%[%(%{]*%s*$', '')
+end
+
+function M.statusline(opts)
   if not parsers.has_parser() then return end
-  local indicator_size = indicator_size or 100
+  local options = opts or {}
+  if type(opts) == 'number' then
+    options = {indicator_size = opts}
+  end
+  local indicator_size = options.indicator_size or 100
+  local type_patterns = options.type_patterns or {'class', 'function', 'method'}
+  local transform_fn = options.transform_fn or transform_line
+  local separator = options.separator or ' -> '
 
   local current_node = ts_utils.get_node_at_cursor()
   if not current_node then return "" end
 
-  local expr = current_node:parent()
-  local prefix = ""
-  if expr then
-    prefix = "->"
-  end
+  local lines = {}
+  local expr = current_node
 
-  local indicator = current_node:type()
-  while expr and (#indicator + #(expr:type()) + 5) < indicator_size do
-    indicator = expr:type() .. prefix .. indicator
+  while expr do
+    local line = get_line_for_node(expr, type_patterns, transform_fn)
+    if line ~= '' and not vim.tbl_contains(lines, line) then
```

Comment:
I don't think we need to check for tbl_contains here

Reply:
It can potentially give some duplicate entries in case when `type_pattern` contains more types. This is just ensure that we are not printing double values in those cases, since we are working with small amount of space.